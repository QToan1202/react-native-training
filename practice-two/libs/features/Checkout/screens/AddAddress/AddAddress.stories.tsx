import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { AddAddress } from '../index'

export default {
  title: 'screens/Checkout/Add Address',
  component: AddAddress,
} as ComponentMeta<typeof AddAddress>

const Template: ComponentStory<typeof AddAddress> = () => <AddAddress />

export const Default = Template.bind({})
Default.args = {}
