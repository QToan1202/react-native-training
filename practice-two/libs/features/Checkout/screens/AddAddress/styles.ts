import { StyleSheet } from 'react-native'

const styles = StyleSheet.create({
  form: {
    marginTop: 36,
    marginHorizontal: 27,
    rowGap: 25,
  },
  error: {
    paddingLeft: 0,
  },
})

export default styles
