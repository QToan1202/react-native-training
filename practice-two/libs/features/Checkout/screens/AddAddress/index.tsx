import { useCallback, useMemo } from 'react'
import { KeyboardAvoidingView, ToastAndroid } from 'react-native'
import { useForm } from 'react-hook-form'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { ScrollView } from 'tamagui'
import { useShallow } from 'zustand/react/shallow'

import { RootStackParamList } from '@navigation/Stack'
import { Button, Input, TabBar } from '@components'
import { IAddressBase, IForm } from '@types'
import { ADDRESS_FORM_FIELDS } from '@constants'
import { TAddressFormFields } from '@constants/screens/addAddressFields'
import { useAddAddress } from '@hooks'
import { useAuthStore } from '@stores'

import styles from './styles'

export type AddAddressScreenProps = NativeStackScreenProps<RootStackParamList, 'AddAddress'>

const AddAddress = ({ navigation }: AddAddressScreenProps) => {
  const { mutate, status } = useAddAddress(process.env.ADDRESS_ENDPOINT)
  const [isHydrated, user] = useAuthStore(useShallow((state) => [state.isHydrated, state.user]))
  const { control, handleSubmit } = useForm<IForm>()
  const handlePress = useCallback(() => undefined, [])
  const handleSaveAddress = useCallback(
    (addressInformation: IAddressBase) => {
      if (!isHydrated) return
      if (!user) return
      // TODO: Save address to context
      mutate(
        {
          ...addressInformation,
          userId: user.id,
        },
        {
          onSuccess: () => {
            navigation.goBack()
            ToastAndroid.show('Add address success!!!', ToastAndroid.SHORT)
          },
        }
      )
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [isHydrated, user]
  )
  const renderFormFields = useMemo(
    () =>
      ADDRESS_FORM_FIELDS.map(({ id, ...fieldData }: TAddressFormFields) => (
        <Input {...fieldData} key={id} control={control} isShowError />
      )),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  )

  return (
    <>
      <ScrollView contentContainerStyle={{ flex: 1 }} backgroundColor="$color.bg_layer">
        <Button
          title="Use current location"
          variant="quaternary"
          color="$color.blue"
          textTransform="none"
          leftIcon={require('@assets/location.png')}
          onPress={handlePress}
        />
        <KeyboardAvoidingView style={styles.form}>{renderFormFields}</KeyboardAvoidingView>
      </ScrollView>
      <TabBar
        title="Save"
        isDisable={status === 'pending'}
        onPress={handleSubmit(handleSaveAddress)}
      />
    </>
  )
}

export default AddAddress
