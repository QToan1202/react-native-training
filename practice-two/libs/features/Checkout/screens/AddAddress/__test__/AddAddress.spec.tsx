import { render } from '@testing-library/react-native'

import AddAddress from '../index'

describe('Testing Add Address screen', () => {
  it('Render screen successfully', () => {
    const component = render(<AddAddress />).toJSON()

    expect(component).toMatchSnapshot()
  })
})
