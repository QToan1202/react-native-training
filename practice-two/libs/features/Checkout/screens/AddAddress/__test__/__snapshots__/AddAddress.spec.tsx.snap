// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Add Address screen Render screen successfully 1`] = `
[
  <RCTScrollView
    contentContainerStyle={
      {
        "backgroundColor": "#f6f9ff",
        "flex": 1,
      }
    }
  >
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#fff",
            "borderColor": "#fff",
            "borderRadius": 0,
            "borderWidth": 2,
            "opacity": 1,
            "paddingVertical": 13,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "columnGap": 6,
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Image
            source={
              {
                "testUri": "../../../assets/location.png",
              }
            }
            style={
              {
                "height": 24,
                "width": 24,
              }
            }
          />
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontFamily": "Montserrat_400Regular",
                },
                {
                  "fontSize": 16,
                },
                [
                  {
                    "color": "#000",
                  },
                  {
                    "fontFamily": "Montserrat_500Medium",
                    "textAlign": "center",
                    "textTransform": "capitalize",
                  },
                  {
                    "color": "#4ea0ff",
                    "textTransform": "none",
                  },
                ],
              ]
            }
          >
            Use current location
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          {
            "marginHorizontal": 27,
            "marginTop": 36,
            "rowGap": 6,
          }
        }
      >
        <View>
          <View>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#3a2c3c",
                    "fontFamily": "Montserrat_500Medium",
                  },
                ]
              }
            >
              Name
            </Text>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              style={
                [
                  {
                    "borderBottomColor": "#dbdbde",
                    "borderBottomWidth": 0.5,
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "fontSize": 16,
                    "paddingVertical": 5,
                  },
                  {},
                  undefined,
                ]
              }
              value=""
            />
          </View>
          <View
            style={
              {
                "height": 20,
              }
            }
          />
        </View>
        <View>
          <View>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#3a2c3c",
                    "fontFamily": "Montserrat_500Medium",
                  },
                ]
              }
            >
              Phone
            </Text>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              style={
                [
                  {
                    "borderBottomColor": "#dbdbde",
                    "borderBottomWidth": 0.5,
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "fontSize": 16,
                    "paddingVertical": 5,
                  },
                  {},
                  undefined,
                ]
              }
              value=""
            />
          </View>
          <View
            style={
              {
                "height": 20,
              }
            }
          />
        </View>
        <View>
          <View>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#3a2c3c",
                    "fontFamily": "Montserrat_500Medium",
                  },
                ]
              }
            >
              Street address
            </Text>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              style={
                [
                  {
                    "borderBottomColor": "#dbdbde",
                    "borderBottomWidth": 0.5,
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "fontSize": 16,
                    "paddingVertical": 5,
                  },
                  {},
                  undefined,
                ]
              }
              value=""
            />
          </View>
          <View
            style={
              {
                "height": 20,
              }
            }
          />
        </View>
        <View>
          <View>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#3a2c3c",
                    "fontFamily": "Montserrat_500Medium",
                  },
                ]
              }
            >
              State
            </Text>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              style={
                [
                  {
                    "borderBottomColor": "#dbdbde",
                    "borderBottomWidth": 0.5,
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "fontSize": 16,
                    "paddingVertical": 5,
                  },
                  {},
                  undefined,
                ]
              }
              value=""
            />
          </View>
          <View
            style={
              {
                "height": 20,
              }
            }
          />
        </View>
        <View>
          <View>
            <Text
              style={
                [
                  {
                    "color": "#fff",
                    "fontFamily": "Montserrat_400Regular",
                  },
                  {
                    "fontSize": 14,
                  },
                  {
                    "color": "#3a2c3c",
                    "fontFamily": "Montserrat_500Medium",
                  },
                ]
              }
            >
              Zipcode
            </Text>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              style={
                [
                  {
                    "borderBottomColor": "#dbdbde",
                    "borderBottomWidth": 0.5,
                    "color": "#000",
                    "fontFamily": "Montserrat_500Medium",
                    "fontSize": 16,
                    "paddingVertical": 5,
                  },
                  {},
                  undefined,
                ]
              }
              value=""
            />
          </View>
          <View
            style={
              {
                "height": 20,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>,
  <View
    style={
      [
        {
          "paddingHorizontal": 16,
          "paddingVertical": 22,
        },
        {
          "backgroundColor": "#33907C",
          "rowGap": 24,
        },
        {
          "paddingTop": 0,
        },
        [
          {
            "backgroundColor": "#fff",
            "paddingHorizontal": 32,
          },
          {
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 12,
              "width": 0,
            },
            "shadowOpacity": 0.6,
            "shadowRadius": 16,
          },
          {
            "paddingBottom": 12,
            "paddingTop": 12,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "columnGap": 4,
            "flexDirection": "row",
          },
          {
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#33907C",
            "borderColor": "#33907C",
            "borderRadius": 50,
            "borderWidth": 2,
            "flex": 1,
            "opacity": 1,
            "paddingVertical": 13,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "columnGap": 6,
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontFamily": "Montserrat_400Regular",
                },
                {
                  "fontSize": 16,
                },
                [
                  {
                    "color": "#fff",
                  },
                  {
                    "fontFamily": "Montserrat_500Medium",
                    "textAlign": "center",
                    "textTransform": "capitalize",
                  },
                  undefined,
                ],
              ]
            }
          >
            Save
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
